security/s2a/internal/crypter/chachapoly_test.go

```diff
@@ -0,0 +1,214 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package crypter
+
+import (
+	"fmt"
+	"testing"
+
+	"google.golang.org/grpc/security/s2a/internal/crypter/testutil"
+)
+
+// getChachaPolyCrypterPair outputs a sender/receiver pair on CHACHA-POLY.
+func getChachaPolyCrypterPair(key []byte, t *testing.T) (s2aAeadCrypter, s2aAeadCrypter) {
+	sender, err := newChachaPoly(key)
+	if err != nil {
+		t.Fatalf("newChachaPoly(ClientSide, key) = %v", err)
+	}
+	receiver, err := newChachaPoly(key)
+	if err != nil {
+		t.Fatalf("newChachaPoly(ServerSide, key) = %v", err)
+	}
+	return sender, receiver
+}
+
+func wycheProofTestVectorFilterChachaPoly(testGroup testutil.TestGroup) bool {
+	return testGroup.IVSize != 96 ||
+		(testGroup.KeySize != 256) ||
+		testGroup.TagSize != 128
+}
+
+func testChachaPolyEncryptionDecryption(sender s2aAeadCrypter, receiver s2aAeadCrypter, test *testutil.CryptoTestVector, t *testing.T) {
+	// Ciphertext is: encrypted text + tag.
+	ciphertext := append(test.Ciphertext, test.Tag...)
+
+	// Encrypt.
+	var dst []byte
+	if test.AllocateDst {
+		dst = make([]byte, len(test.Plaintext)+sender.tagSize())
+	}
+	got, err := sender.encrypt(dst[:0], test.Plaintext, test.Nonce, test.Aad)
+	if testutil.IsFailure(test.Result, err, got, ciphertext) {
+		t.Errorf("key=%v\nEncrypt(\n dst = %v\n plaintext = %v\n nonce = %v\n aad = %v\n) = (\n %v\n %v\n), want %v",
+			test.Key, dst[:0], test.Plaintext, test.Nonce, test.Aad, got, err, ciphertext)
+	}
+
+	// Decrypt.
+	got, err = receiver.decrypt(nil, ciphertext, test.Nonce, test.Aad)
+	if testutil.IsFailure(test.Result, err, got, test.Plaintext) {
+		t.Errorf("key=%v\nDecrypt(\n dst = nil\n ciphertext = %v\n nonce = %v\n aad = %v\n) = (\n %v\n %v\n), want %v",
+			test.Key, ciphertext, test.Nonce, test.Aad, got, err, test.Plaintext)
+	}
+}
+
+func testChachaPolyEncryptRoundtrip(sender s2aAeadCrypter, receiver s2aAeadCrypter, t *testing.T) {
+	// Construct a dummy nonce.
+	nonce := make([]byte, nonceSize)
+
+	// Encrypt.
+	const plaintext = "This is plaintext."
```

Comment:
Is there any test where `plaintext` is empty? E.g. does one of the Wycheproof test vectors cover this case?

Reply:
Can confirm there are multiple test cases where the plaintext is empty, in combination with other edge cases.

Reply:
SG. Thanks! :)